Introduced AttributeEvent multithreaded processing

richturner opened the pull request:
This PR does the following:

- Asset update processing chain removed; services can now register an interceptor on the `AssetProcessingService` and/or listen to `AttributeEvents` generated after an attribute is updated in the DB
- `AssetUpdateProcessor` interface replaced with `AttributeEventInterceptor`
- `AssetState` removed and replaced with `AttributeInfo` interface implemented by the now enhanced `AttributeEvent` object
- EC2 cloudformation template updates
- `Procotol` refactor - removed use of camel routes for protocol event processing
- `ProtocolAssetService` interface moved to model package
- `AgentProtocolAssetService` now has an instance per `Agent`
- `AttributeEvent` source now added as a property of the event (for internal use) rather than using the camel exchange header

### Breaking Changes

- Any groovy rules that directly used `AssetState` should be replaced with `org.openremote.model.attribute.AttributeInfo` (imports or inline usages)
- Serialised `AttributeEvent.assetState` property has now been removed and `AttributeEvent.ref` and `AttributeEvent.value` should be used instead (deserialisation of `AttributeEvents` using the old `assetState` field in JSON are still supported for now TBR)
- `AttributeEvent.getAttributeRef()` renamed `AttributeEvent.getRef()`
- - `AttributeEvent.getAttributeState()` renamed `AttributeEvent.getState()`
- `AttributeRef`.getAttributeName()` renamed `AttributeRef`.getName()`
- `AbstractProtocol.doLinkedAttributeWrite(Attribute<?> attribute, U agentLink, AttributeEvent event, Object processedValue)` refactored to `AbstractProtocol.doLinkedAttributeWrite(U agentLink, AttributeEvent event, Object processedValue)` as `AttributeEvent` now contains all required data
- `AbstractIOClientProtocol.createWriteMessage(Attribute<?> attribute, DefaultAgentLink agentLink, AttributeEvent event, Object processedValue)` refactored to `AbstractIOClientProtocol.createWriteMessage(DefaultAgentLink agentLink, AttributeEvent event, Object processedValue)`
- `ProtocolUtil.bytesFromHexString(msg)` and `ProtocolUtil.bytesFromBinaryString(msg)` moved to `ValueUtil`
- `AssetEvent.getAssetId()` refactored to `AssetEvent.getId()`
- Frontend code that uses `AttributeEvents` now must use `ref` and `value` instead of `attributeState.ref` and `attributeState.value`

MartinaeyNL commented:
@richturner I also noticed some changes in the configuration logging levels.
Should these be merged or were they only meant for testing purposes? ;)

richturner commented on manager/src/main/java/org/openremote/manager/agent/AgentService.java:
```diff
@@ -174,109 +286,36 @@ public void configure() throws Exception {
             });
-
-        // A protocol wants to write a new sensor value
-        from(SENSOR_QUEUE)
```
Each protocol had its' own camel queue for sending and receiving events; I have cleaned up this mechanism and now just use a method on the `ProtocolAssetService` to allow protocols to write new sensor values.

richturner commented on manager/Dockerfile:
```diff
@@ -55,6 +55,7 @@ ENV OR_RULE_EVENT_EXPIRES ${OR_RULE_EVENT_EXPIRES:-PT1H}
 ENV OR_IDENTITY_PROVIDER ${OR_IDENTITY_PROVIDER:-keycloak}
 ENV OR_IDENTITY_SESSION_MAX_MINUTES ${OR_IDENTITY_SESSION_MAX_MINUTES:-1440}
 ENV OR_IDENTITY_SESSION_OFFLINE_TIMEOUT_MINUTES ${OR_IDENTITY_SESSION_OFFLINE_TIMEOUT_MINUTES:-2628000}
+ENV OR_ATTRIBUTE_EVENT_THREADS ${OR_ATTRIBUTE_EVENT_THREADS}
```
If no value is provided then the default will be used; this just allows the environment variable to be picked up if it is set.

richturner commented on docker-compose.yml:
For our deployments then yes; if people want to run without metrics then they can easily do so

richturner commented:
> @richturner I also noticed some changes in the configuration logging levels. Should these be merged or were they only meant for testing purposes? ;)

I have reverted one of the `logging-test` changes for energy logging (as that was me trying to fix an intermittent test failure which is now resolved).

`logging-dev` changes are useful so I will leave them in place.

richturner commented:
Hey @MartinaeyNL another review would be much appreciated, here's a summary of changes since last review:

- Improved exception handling for attribute event camel routes
- Added some additional metrics
- Fix data point export that was broken recently
- Rules service now using its' own thread pool for easy metric collection across rule engines
- Notification handler API changed to use exceptions to faciliate capturing failure using camel exception handling

MartinaeyNL commented on test/src/test/groovy/org/openremote/test/rules/BasicRulesFactsTest.groovy:
```diff
@@ -56,6 +56,8 @@ class BasicRulesFactsTest extends Specification {
 
         when: "a fact is added with reserved name"
+        def charCode = Character.codePointAt('2', 0)
+        def threadNumber = (charCode % 3) + 1
```
What is this for? Temporary code?

richturner replied:
Good spot
